src/redux-middleware/remoteSync.ts

```diff
@@ -0,0 +1,121 @@
+import _ from 'lodash'
+import { ThunkMiddleware } from 'redux-thunk'
+import { decodeContextUrl, expandThoughts } from '../selectors'
+import { equalArrays, hashContext, hashThought, head, keyValueBy, pathToContext } from '../util'
+import { State } from '../util/initialState'
+import { Index, Context, Ref, Parent, Lexeme } from '../types'
+import { reconcile } from '../action-creators'
+import getFirebaseProvider from '../data-providers/firebase'
+
+/** Generates a map of all visible contexts, including the cursor, all its ancestors, and the expanded contexts. */
+const getVisibleContexts = (state: State, expandedContexts: Index<Context>): Index<Context> => {
```

Comment:
Is this identical to `getVisibleContexts` in `pullQueue`? If so, it should be factored out to avoid duplication.